Add widgets to create a wizard app.

csadorf opened the pull request:
Resolves #166.

yakutovicha commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,183 @@
+import ipywidgets as ipw
+
+
+class WizardApp(ipw.VBox):
```
```suggestion
class WizardAppWidget(ipw.VBox):
```

Since it is a widget, I would propose to rename it like this. Not all the existing widgets are renamed like this, but I would suggest we use the `...Widget` naming convention.

csadorf replied:
Done.

yakutovicha commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,183 @@
+
+
+class WizardApp(ipw.VBox):
+    class State(Enum):
```
Would it make sense to move this guy to `WizardAppStep`? This way we don't need to import both `WizardApp` and `WizardAppStep` when creating a new widget.

csadorf replied:
Yes, makes more sense. Done.

yakutovicha reviewed (commented):
Thanks, @csadorf, please see my comments below.

Additionally, I would ask you to create a small notebook example demonstrating how to use it. Maybe use the code I used for testing: 
```python

import ipywidgets as ipw
from aiidalab_widgets_base import WizardApp, WizardAppStep

class MyClass(ipw.VBox, WizardAppStep):
    def __init__(self):
        self.state = WizardApp.State.INIT
        button = ipw.Button(description="Complete")
        super().__init__([button])
    def _on_complete(self, _=None):
        self.state = WizardApp.State.SUCCESS

w = WizardApp(steps=[("title1", MyClass()), ("title2", MyClass()), ("title3", MyClass())])

display(w)
```
I am not sure if this is exactly how things are supposed to work, maybe I just miss something.

csadorf commented:
> Thanks, @csadorf, please see my comments below.
> 
> Additionally, I would ask you to create a small notebook example demonstrating how to use it (which actually didn't work as I expected). Maybe for the starting point use the code I wrote for testing:
> 
> ```python
> import ipywidgets as ipw
> from aiidalab_widgets_base import WizardApp, WizardAppStep
> 
> class MyClass(ipw.VBox, WizardAppStep):
>     def __init__(self):
>         self.state = WizardApp.State.INIT
>         button = ipw.Button(description="Complete")
>         super().__init__([button])
>     def _on_complete(self, _=None):
>         self.state = WizardApp.State.SUCCESS
> 
> w = WizardApp(steps=[("title1", MyClass()), ("title2", MyClass()), ("title3", MyClass())])
> 
> display(w)
> ```
> 
> I am not sure if this is exactly how things are supposed to work, maybe I just miss something.

@yakutovicha I've pushed a "pizza order" apps as an example within the `wizard_apps.ipynb` notebook. I tried to keep the example concise, but demonstrate all relevant components and how they are supposed to work together. Unfortunately the "minimal" example is more elaborate than I hoped it to be, but I am not sure how to significantly shorten this without sacrificing the actual demonstration. I'd appreciate if you could review this before I address the other change requests you made.

yakutovicha commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,183 @@
+
+    ICONS = {
+        State.INIT: "\u25cb",
+        State.CONFIGURED: "\u25ce",
+        State.READY: "\u25ce",
```
```suggestion
        State.READY: "\u25ce",
        State.CONFIGURED: "\u25cf",
```

What I understood from your example is that the `CONFIGURED` state is more "complete".

I prefer the following symbols:

`INIT` - empty circle, ready to be filled in. [as it is now]
`READY` - two concentric circles, in the process of filling [as it is now]
`CONFIGURED` - full circle, everything is provided [taken from the `SUCCESS`]
`ACTIVE` - rotating circle [ as it is now]
`SUCCESS` - check mark
`FAILURE` - cross

csadorf replied:
Done.

yakutovicha reviewed (commented):
Thanks, @csadorf.

Some further suggestions after reviewing your example notebook.

Example:
1. I would propose to have 3 steps: Configure, Review order, Delivery (with a bar that follows pizza delivery).
2. Add a failure case, like "Pizza you configured is not available".
3. Add a reset functionality.

Widget:
1. Slightly different icons (see my suggestion below)
2. Move back button should be enabled if the previous step isn't completed.

csadorf commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,183 @@
+        State.CONFIGURED: "\u25ce",
+        State.READY: "\u25ce",
+        State.ACTIVE: "\u25d4",
+        State.SUCCESS: "\u25cf",
+        State.FAIL: "\u25cd",
```
Done.

csadorf commented:
>     1. I would propose to have 3 steps: Configure, Review order, Delivery (with a bar that follows pizza delivery).

Done, except for the progress bar (this is already getting out of hand in complexity).

>     2. Add a failure case, like "Pizza you configured is not available".

Done.

>     3. Add a reset functionality.
> 
Done; also adjusted the function to reset the whole app instead of individual steps. The previous logic (reset based on step) was very hard to get right and sometimes downright impossible. For the sake of transparency and ensuring that app developers can implement something that reliably works, I've adjusted this step to mean "reset the whole app". There are reasonable defaults implemented as to when that should be possible, but app developers at the very least need to implement a step `reset()` function.

> Widget:
> 
>     1. Slightly different icons (see my suggestion below)

Done.

>     2. Move back button should be enabled if the previous step isn't completed.

Done. Also required a bit of an adjustment in overall logic (please see the commit message for details).

csadorf commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,183 @@
+        )
+        self.next_button.on_click(self._on_click_next_button)
+
+        self.footer = ipw.HBox(
```
Good catch! Done.

csadorf commented:
@yakutovicha I've applied all the other remaining change requests.

csadorf commented:
@CasperWA Maybe also have a look at the pizza order example and let me know if anything is unclear. All concepts needed should be demonstrated there, but obviously not every app needs all of these functions.

yakutovicha reviewed (commented):
Thanks, @csadorf, I put some comments below to further improve the wizard. I think, once they are addressed, the widget is practically done.

Concerning the example, I have some more ideas. I think `ReviewAndSubmitOrderStep` and `TrackOrderStep` are too much interdependent. I would reduce the interdependency to simplify things down. Here is what I would do:

1. I would prefer that "pizza unavailable" triggers the "Order Pizza" step to fail. I think the failure should happen right after you submit the order. This way we can demonstrate that going to step 3 is impossible if step 2 had failed.
2. I would move all the delivery-related stuff to the `TrackOrderStep` class. This involves `order_status` enum, `process_order` function.
3. I would rename "Order Pizza" to "Confirm the Order", but this is a minor thing. Up to you.

csadorf commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,212 @@
+
+        # Watch for changes to each step's state
+        for widget in widgets:
+            assert widget.has_trait("state")
```
Explicit type checks are not in the spirit of Python's duck-typing (dynamic typing) paradigm and unless really needed should be avoided as it poses an unnecessary constraint.

CasperWA replied:
Pragmatism vs. idealism :)
In any case, `assert` shouldn't be used here, as it's considered a statement for debugging and testing, e.g. all `assert` statements can be subverted running the python interpreter with the `-O` option.

yakutovicha replied:
Ok, then let's look at this from a slightly different angle. To make sure things work properly, do we want to require that widgets are derived from `WizardAppWidgetStep`? Or should we allow widgets that are not derived from it?

csadorf replied:
We are explicitly observing that `state` trait in the next line it would fail either way (even with `-O` option...), the assertion could just as well be removed entirely. I don't really see a point why one would not use a `WizardAppWidgetStep` to implement a step, but I also see no real point in making that a hard requirement. To make sure that apps work, developers must test them. If they forget to implement the `step` interface in their steps, the app is going to fail. You could use the same argument for or against type checks in the rest of the language.

yakutovicha replied:
> To make sure that apps work, developers must test them. If they forget to implement the `step` interface in their steps, the app is going to fail.

My advice on the explicit error was actually to help developers in writing their widgets. Additionally, requiring the widgets to derive from `WizardAppWidgetStep` would help in ensuring backwards compatibility. Imagine the relations between `WizardAppWidget` - `WizardAppWidgetStep` change. If a developer wasn't using the `WizardAppWidgetStep` as a base class - things might not be working anymore for them.

csadorf replied:
I understand that, but the documentation is helping developers already with that. We are showing that you should derive your steps from this class and things are going to immediately fail if a developers forgets to do so. I could provide a more detailed error message here, something like:

> The step widget has no `.state` trait, did you forget to derive the step from the `WizardAppWidgetStep` class?

However, I just don't see a good reason to stray from core language paradigms here.

> Additionally, requiring the widgets to derive from WizardAppWidgetStep would help in ensuring backwards compatibility. That does not really have anything to do with this problem IMO.

Again, developers should absolute derive from that class in 99% of cases – this is not in question. I just don't really a good reason to *force* them to.

csadorf commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,212 @@
+                self.reset_button.disabled = not self.can_reset()
+
+    def reset(self, step=0):
+        """Reset the app down to the given step.
```
Point 1: Good catch, I will fix the doc-string.

Point 2: You mean start from the last step and up? Otherwise we would end up in an inconsistent state. In principle, I agree, however we use `hold_sync()` to actually prevent any synchronization between steps during the reset, so we generate a given state and then only after the hold is lifted we perform the consistency checks.

yakutovicha replied:
> You mean start from the last step and up?

Yes.

>  In principle, I agree, however we use hold_sync() to actually prevent any synchronization between steps during the reset, so we generate a given state and then only after the hold is lifted we perform the consistency checks.


Ok, true. Then it is up to you if you want to reverse the order.

csadorf commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,212 @@
+        )
+        self.next_button.on_click(self._on_click_next_button)
+
+        self.header = ipw.HBox(
```
Done.

csadorf commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,212 @@
+    def __init__(self, steps, **kwargs):
+        # The number of steps must be greater than one
+        # for this app's logic to make sense.
+        assert len(steps) > 1
```
Done.

csadorf commented on aiidalab_widgets_base/wizard.py:
```diff
@@ -0,0 +1,212 @@
+            icon = self.icons().get(widget.state, str(widget.state).upper())
+            self.accordion.set_title(i, f"{icon} Step {i+1}: {title}")
+
+    def _consider_switch(self, _=None):
```
Done.

csadorf commented:
>     1. I would prefer that "pizza unavailable" triggers the "Order Pizza" step to fail. I think the failure should happen right after you submit the order. This way we can demonstrate that going to step 3 is impossible if step 2 had failed.

I think that example is worse. We want to demonstrate how to handle runtime issues (the kind we experience with simulations etc.) If we could check for the availability of ingredients before submission – and thus have knowledge about the availability – then that particular ingredient should be disabled from the start.

>     2. I would move all the delivery-related stuff to the `TrackOrderStep` class. This involves `order_status` enum, `process_order` function.

Why? Then we need to replicate the configuration trait again. The idea is that we submit something in step 2 (like a simulation) and then track its state in step 3. That's like handing over a process node that has the state information. Is the problem with starting the thread in step 2?

>     3. I would rename "Order Pizza" to "Confirm the Order", but this is a minor thing. Up to you.

Sure.

csadorf commented:
@yakutovicha I could create an 'Order' class if you like that would represent a direct 1-to-1 example for how to handle submitted processes. Would that help to make things clearer?

yakutovicha commented:
> @yakutovicha I could create an 'Order' class if you like that would represent a direct 1-to-1 example for how to handle submitted processes. Would that help to make things clearer?

Let me just quickly play with the example. I will try to understand whether my suggestions indeed improve the clearness and whether I can address [your comments](https://github.com/aiidalab/aiidalab-widgets-base/pull/180#issuecomment-799378878). In case I manage, I will put my example notebook here in the comments.

yakutovicha commented:
> > @yakutovicha I could create an 'Order' class if you like that would represent a direct 1-to-1 example for how to handle submitted processes. Would that help to make things clearer?
> 
> Let me just quickly play with the example:

[Here it is](https://github.com/aiidalab/aiidalab-widgets-base/files/6142578/wizard_apps_my.ipynb.txt), please have a look. The reset functionality is not working - I didn't fix that as this is just to demonstrate what I meant.

> Why? Then we need to replicate the configuration trait again

Actually, I didn't need that. 

> I think that example is worse. We want to demonstrate how to handle runtime issues (the kind we experience with simulations etc.) 

Actually, I just provided another reason to fail (no delivery on Monday) for the delivery.

yakutovicha commented:
> If that is so, then I don't think the additional complexity is warranted.

well, for me, the additional complexity is the `ConfirmOrder` class that has delivery info. I really got confused the first time I saw the code and I believe the others will be confused too. If the price for clearness would be having the `configuration` trait available in all 3 classes, I would pay this price.

What concerns the failure modes, I believe we should cover two cases:
1. fail-during-setup
```
[done]
[failed]
[unavailalbe]
```
2. fail-during-execution
```
[done]
[done]
[failed]
```

> because it clearly demonstrates an "exception" case that could be difficult to anticipate during the order process

ok, I agree that your failure case is a good example, honestly, I can't think of a better one.


**An idea for the fail-during-setup**: "cancel order" button to the confirmation step that would put it into the failure mode. What do you think?

csadorf commented:
> > If that is so, then I don't think the additional complexity is warranted.
> 
> well, for me, the additional complexity is the `ConfirmOrder` class that has delivery info. I really got confused the first time I saw the code and I believe the others will be confused too. If the price for clearness would be having the `configuration` trait available in all 3 classes, I would pay this price.

I am ok with refactoring the order simulation logic into a separate class. I already offered that before.

> What concerns the failure modes, I believe we should cover two cases:
> 
>     1. fail-during-setup

I think we have such a hard time getting on the same page here, because you seem to have 2 misconceptions.

1. This is not a valid failure state for this example workflow. The `FAIL` state is meant to be used for unrecoverable **runtime** failures. For example, if it was possible to detect that certain ingredients are not available prior to submission, then the interface should prevent the submission in the first place instead of entering a failure state. Since we are not running anything in step 2, there is no valid `FAIL` state.

2. But even if there was one, it would still make no sense, because there is nothing special about the `FAIL` state in comparison to any other state when it comes to enabling the next step. The next step is only available if the previous step is in the `SUCCESS` state. **This means "preventing" the enabling of the next step is already demonstrated!**

> ok, I agree that your failure case is a good example, honestly, I can't think of a better one.
> 
> **An idea for the fail-during-setup**: "cancel order" button to the confirmation step that would put it into the failure mode. What do you think?

I do not think that makes sense due to the reasons I explained in the previous paragraph.

csadorf commented:
@yakutovicha This is ready for your final review.

yakutovicha commented on wizard_apps.ipynb:
```diff
@@ -0,0 +1,325 @@
+    "\n",
+    "            # Link the order status to the widget displaying it to the user:\n",
+    "            ipw.dlink((change[\"new\"], \"status\"), (self.status_widget, \"status\"))\n",
+    "\n",
```
```suggestion
    "        else:\n",
    "            self.state = self.State.INIT\n",
    "\n",
```

Otherwise the third step fails to reset.

csadorf replied:
Thx a lot! 😄

yakutovicha reviewed (changes requested):
Thanks, @csadorf, looks good to me.

I found a little bug: the third step wasn't properly reset (see my suggestion below). Once this is fixed, I am happy to approve the PR.

csadorf commented:
@yakutovicha I've further extended the documentation and fixed one more bug (the first step should be in READY state by default, not INIT state).

yakutovicha reviewed (approved):
Great work, thanks a lot @csadorf!

I let you merge it.
